### ISSUE TITLE
Pickle Support

### ISSUE THREAD

**Issue by SerialDev** [PyO3/pyo3#100]
As of right now its not possible to pickle classes created by PyO3.  
This feature would be invaluable for situations where some form of persistence would be desireable.

**Comment by rth** [comment]
Would this mean implementing `__getstate__` and `__setstate__` methods (cf https://docs.python.org/3/library/pickle.html#pickling-class-instances)?

For instance, the way pickling works for,
 - [Boost.Python C++ library](https://www.boost.org/doc/libs/1_58_0/libs/python/doc/v2/pickle.html)
 - [Cython extensions](https://cython.readthedocs.io/en/latest/src/userguide/extension_types.html#controlling-pickling)

might provide some examples.

**Comment by rth** [comment]
For instance, if we take the documentation example for  `MyClass`,

<details>

```rust
# use pyo3::prelude::*;
# use pyo3::PyRawObject;
#[pyclass]
struct MyClass {
   num: i32,
}

#[pymethods]
impl MyClass {

     #[new]
     fn new(obj: &PyRawObject, num: i32) {
         obj.init({
             MyClass {
                 num,
             }
         });
     }
}
```
</details>

by default we get the following error when pickling this class,
```python
        obj = MyClass()
    
>       pickle.dumps(obj)
E       TypeError: can't pickle MyClass objects
```

If we now add the `__getstate__`/ `__setstate__` methods,
```rust
    fn __getstate__(&self) -> PyResult<(i32)> {
        Ok(self.num)
    }

    fn __setstate__(&mut self, state: i32) -> PyResult<()> {
        self.num = state;
        Ok(())
    }
```
we get another exception,
```
_pickle.PicklingError: Can't pickle <class 'MyClass'>: attribute lookup MyClass on builtins failed
```
There some step I must be missing here.

**Comment by althonos** [comment]
@rth : this may be related to the fact that PyO3 exposes all classes as part of the `builtins` module, because the import mechanism has not been properly implemented, so pickle tries to use `builtins.MyClass` and fails with the error you reported.

**Comment by rth** [comment]
Thanks @althonos ! Opened a separate issue about it in #474

**Comment by rth** [comment]
So by subclassing , to set the `__module__` correctly as suggested in https://github.com/PyO3/pyo3/issues/474#issuecomment-489521285, pickling seems to work. 

Though, I get a segfault occasionally (i.e. it does seem to be random) at exit. For instance when running a pytest session where one test checks pickling,
```
gdb --args python3.7 -m pytest -k test_pickle
GNU gdb (GDB) CentOS (7.0.1-45.el5.centos)
[...]
Reading symbols from /opt/_internal/cpython-3.7.1/bin/python3.7...(no debugging symbols found)...done.
(gdb) run
Starting program: /opt/_internal/cpython-3.7.1/bin/python3.7 -m pytest -k test_pickle
warning: Error disabling address space randomization: Operation not permitted
============================================================= test session starts =============================================================
platform linux -- Python 3.7.1, pytest-4.4.1, py-1.8.0, pluggy-0.9.0 -- /opt/_internal/cpython-3.7.1/bin/python3.7
cachedir: .pytest_cache
rootdir: /src/python
collected 1 items  / 1 selected                                                                                               

my_module/test_pickle.py::test_pickle PASSED

=================================================== 1 passed  in 0.13 seconds ===================================================
During startup program terminated with signal SIGSEGV, Segmentation fault.
(gdb) bt
No stack.
```
and there is no backtrace.

**Comment by konstin** [comment]
The segfault likely occurs because subclassing is broken

**Comment by gilescope** [comment]
How about trying dill? Pickle can't handle lots of pure python serialisation cases.
https://pypi.org/project/dill/

**Comment by davidhewitt** [comment]
Not sure if it's interesting; this snippet just got shared on gitter. https://gist.github.com/ethanhs/fd4123487974c91c7e5960acc9aa2a77

**Comment by shaolo1** [comment]
I've got a simple struct that I need to deepcopy.  I'm trying to figure out how to pickle my struct (after getting the TypeError: cannot pickle error).  The gist above shows how to do it for a single member, but I'm too much of a newb to see how to do this with multiple members.

I tried 
```
pub fn __getstate__(&self, py: Python) -> PyResult<PyObject> {
        Ok(PyBytes::new(py, &serialize(&self.foo).unwrap()).to_object(py))
        Ok(PyBytes::new(py, &serialize(&self.bar).unwrap()).to_object(py))
    }

```

..but get an error "expected one of `.`, `;`, `?`, `}`, or an operator" after the first OK.

**Comment by davidhewitt** [comment]
@shaolo1 I would just return the tuple of members:

```
    pub fn __getstate__(&self, py: Python) -> PyObject {
        (
            PyBytes::new(py, &serialize(&self.foo)?),
            PyBytes::new(py, &serialize(&self.bar)?),
        ).to_object(py)
    }
```

**Comment by shaolo1** [comment]
@davidhewitt Thanks.  I'll try that if I encounter it again.  I got around the problem by just implementing __deepcopy__ in the parent object and handling the copy there so that pickle support was not needed in my rust object.

**Comment by Genarito** [comment]
Was anyone able to solve this for a structure with multiple fields? If so, could you please share the code?

**Comment by frs17** [comment]
I am just getting started with py03 (really impressed!) and am running into this issue. In Python I have a working
`lst = [MyClass() for i in range(100)]` where MyClass is a `#[pyclass]` written in Rust. When I try to speed up the generation of this list via `pathos.ProcessPool` I run into trouble with the Pickling. 
```
from pathos.pools import ProcessPool
def new_class(): 
    return MyClass( )

pool = ProcessPool()
lst2 = pool.map(new_class, range(100))
```
```multiprocess.pool.MaybeEncodingError: Error sending result: ... Reason: 'TypeError("cannot pickle```

In theory pathos.ProcessPool uses `dill` under the hood. I would be grateful to any advice on how I might accomplish my goal here?

**Comment by kylecarow** [comment]
I was able to enable pickling by writing the `__getstate__`, `__setstate__`, and `__getnewargs__` magic methods in pymethods for a pure Rust project using `bincode::{deserialize, serialize}`. In `__getnewargs__` you need to return a tuple of all the arguments `__new__` will use on deserializaton, otherwise you'll see something like `TypeError: MyStruct.__new__() missing 2 required positional arguments: 'my_first_arg' and 'my_second_arg'`.

Here is a generic example:
``` rust
pub fn __setstate__(&mut self, state: Vec<u8>) -> PyResult<()> {
    *self = deserialize(&state).unwrap();
    Ok(())
}
pub fn __getstate__(&self) -> PyResult<Vec<u8>> {
    Ok(serialize(self).unwrap())
}
pub fn __getnewargs__(&self) -> PyResult<(f64, f64)> {
    Ok((self.my_first_arg, self.my_second_arg))
}
```

Also, here is a code example for the workaround @shaolo1 mentioned. Cloning for deepcopy may be faster than serializing & deserializing (which I guess is how Python deepcopies normally?), but I haven't tested that.
``` rust
pub fn copy(&self) -> Self {self.clone()}
pub fn __copy__(&self) -> Self {self.clone()}
pub fn __deepcopy__(&self, _memo: &PyDict) -> Self {self.clone()}
```
That'll allow you to return a clone using `copy.copy()`, `copy.deepcopy()`, or by calling the `.copy()` method.

**Comment by davidhewitt** [comment]
Yes, `Vec<u8>` will cast each byte in turn into a Python list. I think you do need to use `PyBytes` here, and it's irrelevant that you want to avoid the GIL because these are Python methods you're implementing.

I think you want something like this:

```rust
pub fn __setstate__(&mut self, state: &PyBytes) -> PyResult<()> {
    *self = deserialize(state.as_bytes()).unwrap();
    Ok(())
}
pub fn __getstate__<'py>(&self, py: Python<'py>) -> PyResult<&'py PyBytes> {
    Ok(PyBytes::new(py, serialize(&self).unwrap()))
}
pub fn __getnewargs__(&self) -> PyResult<(f64, f64)> {
    Ok((self.my_first_arg, self.my_second_arg))
}
```

I would also strongly recommend you replace `.unwrap()` with conversion to actual `PyResult` errors :)

**Comment by kylecarow** [comment]
Woah, yeah that sped up my round trip serializing and deserializing benchmark by 100x. And thanks for the tip about PyResult errors. I did have to modify `__getstate__` ever so slightly to add a reference:

``` rust
pub fn __getstate__<'py>(&self, py: Python<'py>) -> PyResult<&'py PyBytes> {
    Ok(PyBytes::new(py, &serialize(&self).unwrap()))
}
```

I also did some benchmarking with my structs regarding the performance of cloning vs. roundtrip pickling and bincode serde, which might be useful to someone:
- Having a copy method that calls `.clone()` is by far the fastest way I've found of copying a pyo3 object. My benchmark took **1.38 usec**
- The next best thing is having bincode serde methods, which roundtrip took **15.6 usec** (before the PyBytes change it took 1.28 msec)
  ``` rust
  pub fn to_bincode<'py>(&self, py: Python<'py>) -> PyResult<&'py PyBytes> {
      Ok(PyBytes::new(py, &serialize(&self).unwrap()))
  }
  #[classmethod]
  pub fn from_bincode(_cls: &PyType, encoded: &PyBytes) -> PyResult<Self> {
      Ok(deserialize(encoded.as_bytes()).unwrap())
  }
- The least performant is pickling, as expected. I guess Python has a lot more overhead here. It took **439 usec** roundtrip.

**Comment by lycantropos** [comment]
since `__setstate__` requires a mutable reference is there a possibility to have a `pickle` support for a [`#[pyclass(frozen)]` class](https://pyo3.rs/v0.19.0/class.html#frozen-classes-opting-out-of-interior-mutability)?

**Comment by davidhewitt** [comment]
You can also use `__getnewargs__` or `__getnewargs_ex__`, which is the simplest option if you can pass all your state directly back to `#[new]` when unpickling (I would guess this is true for most `frozen` classes).

**Comment by Stargateur** [comment]
there is no way to avoid construct the object before ? I can't understand why they forced to construct the object and THEN set it to specific state. https://peps.python.org/pep-0307 is very hard to read

**Comment by Stargateur** [comment]
If like me you have trouble to use reduce here a very simple example:

```rust
#[staticmethod]
pub fn deserialize(data: Vec<u8>) -> Self {
    Foo {
        inner: rmp_serde::from_slice(&data).unwrap(),
    }
}

pub fn __reduce__(&self) -> (PyObject, PyObject) {
    Python::with_gil(|py| {
        py.run_bound("import mylib", None, None).unwrap();
        let deserialize = py.eval_bound("mylib.Foo.deserialize", None, None).unwrap();
        let data = rmp_serde::to_vec(&self.inner).unwrap;
        (deserialize.to_object(py), (data,).to_object(py))
    })
}
```

If someone have a way to avoid `run_bound()` and `eval_bound` I'm all hear. Look https://docs.rs/pyo3/0.22.2/pyo3/types/struct.PyCFunction.html#method.new_with_keywords_bound could be used but... don't know how.

**Comment by landaire** [comment]
Hey folks, apologies for the self-promotion but a while ago I forked `serde_pickle` to add some features I needed for a parsing data from a videogame whose data format I was not in control of.

tl;dr of what I've added:

1. Proper refcounted objects
2. Support for globals
3. ...and recursive objects
4. (literally today) support for de/serializing custom classes

I have no idea if this is useful for you all here, but with the headaches I've had from `pickle` I started poking around recently to see if there are others suffering with vinegar-soaked wounds.

Most of this is relatively untested ITW and I'm not a `pickle` expert. I did not use an LLM for any of this _except for today_ I had Claude take a stab though at closing the test gap I had and double-checking the behaviors matched that of upstream pickle.

https://github.com/landaire/pickled

**Comment by denehoffman** [comment]
I'm wondering if there's any interest in making this an opt-in automatic part of the macro iff Serialize and Deserialize are implemented. I was bike-shedding a proc-macro idea for this earlier today, but haven't tried it out in full yet, basically you'd generate something like this:

```rust
const _: fn() = || {
    fn assert_pickle_bounds<T>()
    where
        T: serde::Serialize + serde::de::DeserializeOwned,
    {}

    assert_pickle_bounds::<MyThing>;
};

#[pyo3::pymethods]
impl MyThing {
    #[classmethod]
    #[pyo3(name = "_pyo3_pickle_from_bytes")]
    fn __pyo3_pickle_from_bytes(
        _cls: &pyo3::Bound<'_, pyo3::types::PyType>,
        bytes: &pyo3::Bound<'_, pyo3::types::PyBytes>,
    ) -> pyo3::PyResult<Self> {
        serde_pickle::from_slice(bytes.as_bytes(), Default::default())
            .map_err(|e| pyo3::exceptions::PyValueError::new_err(e.to_string()))
    }

    fn __reduce__<'py>(
        &self,
        py: pyo3::Python<'py>,
    ) -> pyo3::PyResult<pyo3::Py< pyo3::PyAny >> {
        let bytes = serde_pickle::to_vec(self, Default::default())
            .map_err(|e| pyo3::exceptions::PyValueError::new_err(e.to_string()))?;

        let cls = py.get_type_bound::<Self>();
        let reconstruct = cls.getattr("_pyo3_pickle_from_bytes")?;

        let args = pyo3::types::PyTuple::new_bound(
            py,
            [pyo3::types::PyBytes::new_bound(py, &bytes)],
        );

        Ok((reconstruct, args).into_py(py))
    }
}
```

**Comment by davidhewitt** [comment]
Some reasons why I've historically found solving pickling comprehensively challenging:
- Some classes might want opaque bytes serialization like the above, others might want to serialize to a tuple of Python objects like a pure-python pickle implementation would.
- Interaction with `frozen` - some types deserialize with `__setstate__`, others might deserialize with `__new__`.
- Module hierarchy - for unpickling to work the class has to have the correct `__module__`. I think we're better than this than we used to be (esp. with declarative modules), however I still think we haven't solved #759 perfectly seamlessly.

**Comment by denehoffman** [comment]
@davidhewitt ah yes, this does make things more complex, I see
